usersession/client/client.go

```diff
@@ -0,0 +1,264 @@
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ *
+ */
+
+package client
+
+import (
+	"bytes"
+	"context"
+	"encoding/json"
+	"fmt"
+	"io"
+	"io/ioutil"
+	"net"
+	"net/http"
+	"net/url"
+	"path/filepath"
+	"strconv"
+	"sync"
+
+	"github.com/snapcore/snapd/dirs"
+	"github.com/snapcore/snapd/logger"
+)
+
+// dialSessionAgent connects to a user's session agent
+//
+// The host portion of the address is interpreted as the numeric user
+// ID of the target user.
+func dialSessionAgent(network, address string) (net.Conn, error) {
+	host, _, err := net.SplitHostPort(address)
+	if err != nil {
+		return nil, err
+	}
+	socket := filepath.Join(dirs.XdgRuntimeDirBase, host, "snapd-session-agent.socket")
+	return net.Dial("unix", socket)
+}
+
+type Client struct {
+	doer *http.Client
+}
+
+func New() *Client {
+	transport := &http.Transport{Dial: dialSessionAgent, DisableKeepAlives: true}
+	return &Client{
+		doer: &http.Client{Transport: transport},
+	}
+}
+
+type Error struct {
+	Kind    string      `json:"kind"`
+	Value   interface{} `json:"value"`
+	Message string      `json:"message"`
+}
+
+func (e *Error) Error() string {
+	return e.Message
+}
+
+type response struct {
+	// Not from JSON
+	uid        int
+	err        error
+	statusCode int
+
+	Result json.RawMessage `json:"result"`
+	Type   string          `json:"type"`
+}
+
+func (resp *response) checkError() {
+	if resp.Type != "error" {
+		return
+	}
+	var resultErr Error
+	err := json.Unmarshal(resp.Result, &resultErr)
+	if err != nil || resultErr.Message == "" {
+		resp.err = fmt.Errorf("server error: %q", http.StatusText(resp.statusCode))
+	} else {
+		resp.err = &resultErr
+	}
+}
+
+func (client *Client) doMany(ctx context.Context, method, urlpath string, query url.Values, headers map[string]string, body []byte) ([]*response, error) {
+	sockets, err := filepath.Glob(filepath.Join(dirs.XdgRuntimeDirGlob, "snapd-session-agent.socket"))
+	if err != nil {
+		return nil, err
+	}
+	var (
+		wg        sync.WaitGroup
+		mu        sync.Mutex
+		responses []*response
+	)
+	for _, socket := range sockets {
+		wg.Add(1)
+		go func(socket string) {
+			defer wg.Done()
+			uidStr := filepath.Base(filepath.Dir(socket))
+			uid, err := strconv.Atoi(uidStr)
+			if err != nil {
+				logger.Noticef("Socket %q does not appear to be in a valid XDG_RUNTIME_DIR", socket)
+				return
+			}
+			u := url.URL{
+				Scheme:   "http",
+				Host:     uidStr,
+				Path:     urlpath,
+				RawQuery: query.Encode(),
+			}
+			req, err := http.NewRequest(method, u.String(), bytes.NewBuffer(body))
+			if err != nil {
+				logger.Noticef("Failed to create HTTP request: %v", err)
```

Comment:
this shouldn't happen but should probably surface as an error

Reply:
I've rearranged things a bit so this will present as an response error for the uid.  It's not clear how to trigger this code path though.

If we were using Go 1.13's NewRequestWithContext, a nil context would do the trick, but the 1.10 compatible NewRequest + WithContext is a panic rather than an error.